src/Std/Data/DHashMap/Raw.lean

```diff
@@ -0,0 +1,438 @@
+    (by change dite .. = true at h; split at h <;> simp_all)
+
+/-- Tries to retrieve the mapping for the given key, returning `fallback` if no such mapping is
+present. -/
+@[inline] def Const.getD [BEq α] [Hashable α] (m : Raw α (fun _ => β)) (a : α) (fallback : β) : β :=
+  if h : 0 < m.buckets.size then
+    Raw₀.Const.getD ⟨m, h⟩ a fallback
+  else fallback -- will never happen for well-formed inputs
+
+/-- Tries to retrieve the mapping for the given key, panicking if no such mapping is present. -/
+@[inline] def Const.get! [BEq α] [Hashable α] [Inhabited β] (m : Raw α (fun _ => β)) (a : α) : β :=
+  if h : 0 < m.buckets.size then
+    Raw₀.Const.get! ⟨m, h⟩ a
+  else default -- will never happen for well-formed inputs
+
+/-- Equivalent to (but potentially faster than) calling `Const.get?` followed by `insertIfNew`. -/
+@[inline] def Const.getThenInsertIfNew? [BEq α] [Hashable α] (m : Raw α (fun _ => β)) (a : α)
+    (b : β) : Option β × Raw α (fun _ => β) :=
+  if h : 0 < m.buckets.size then
+    let ⟨replaced, ⟨r, _⟩⟩ := Raw₀.Const.getThenInsertIfNew? ⟨m, h⟩ a b
+    ⟨replaced, r⟩
+  else (none, m) -- will never happen for well-formed inputs
+
+end
+
+/-- Returns `true` if the hash map contains no mappings.
+
+Note that if your `BEq` instance is not reflexive or your `Hashable` instance is not
+lawful, then it is possible that this function returns `false` even though is not possible
+to get anything out of the hash map. -/
+@[inline] def isEmpty (m : Raw α β) : Bool :=
+  m.size == 0
+
+section Unverified
+
+/-! We currently do not provide lemmas for the functions below. -/
+
+/-- Updates the values of the hash map by applying the given function to all mappings, keeping
+only those mappings where the function returns `some` value. -/
+@[inline] def filterMap {γ : α → Type w} (f : (a : α) → β a → Option (γ a)) (m : Raw α β) :
+    Raw α γ :=
+  if h : 0 < m.buckets.size then
+    Raw₀.filterMap f ⟨m, h⟩
+  else ∅ -- will never happen for well-formed inputs
+
+/-- Updates the values of the hash map by applying the given function to all mappings. -/
+@[inline] def map {γ : α → Type w} (f : (a : α) → β a → γ a) (m : Raw α β) : Raw α γ :=
+  if h : 0 < m.buckets.size then
+    Raw₀.map f ⟨m, h⟩
+  else ∅ -- will never happen for well-formed inputs
+
+/-- Removes all mappings of the hash map for which the given function returns `false`. -/
+@[inline] def filter (f : (a : α) → β a → Bool) (m : Raw α β) : Raw α β :=
+  if h : 0 < m.buckets.size then
+    Raw₀.filter f ⟨m, h⟩
+  else ∅ -- will never happen for well-formed inputs
+
+/-- Folds the given function over the mappings in the hash map in some order. -/
+@[inline] def foldlM (f : δ → (a : α) → β a → m δ) (init : δ) (b : Raw α β) : m δ :=
+  b.buckets.foldlM (fun acc l => l.foldlM f acc) init
+
+/-- Folds the given function over the mappings in the hash map in some order. -/
+@[inline] def foldl (f : δ → (a : α) → β a → δ) (init : δ) (b : Raw α β) : δ :=
+  Id.run (b.foldlM f init)
```

Comment:
Are these really left folds, given that the order is unspecified? Should they be called `foldM` and `fold`, respectively?

Reply:
Yes, you're right. Not sure what I was thinking.